File: breadth-first-search/breadth_first_search.rb
def bfs(graph, start)
  visited = Set.new
  queue = [start]
  visited.add(start)
  
  while !queue.empty?
    node = queue.shift
    print "#{node} "
    
    (graph[node] || []).each do |neighbor|
      unless visited.include?(neighbor)
        visited.add(neighbor)
        queue.push(neighbor)
      end
    end
  end
end

graph = {0 => [1, 2], 1 => [2], 2 => [0, 3], 3 => [3]}
bfs(graph, 2)

File: breadth-first-search/breadth_first_search.php
<?php
function bfs($graph, $start) {
    $visited = [];
    $queue = [$start];
    $visited[$start] = true;
    
    while (!empty($queue)) {
        $node = array_shift($queue);
        echo $node . " ";
        
        if (isset($graph[$node])) {
            foreach ($graph[$node] as $neighbor) {
                if (!isset($visited[$neighbor])) {
                    $visited[$neighbor] = true;
                    $queue[] = $neighbor;
                }
            }
        }
    }
}

$graph = [
    0 => [1, 2],
    1 => [2],
    2 => [0, 3],
    3 => [3]
];
bfs($graph, 2);
?>